packages/core/src/providers/injectedProvider/provider.tsx

```diff
@@ -0,0 +1,48 @@
+import { ReactNode, useCallback, useEffect, useState } from 'react'
+import { InjectedProviderContext } from './context'
+import { Web3Provider } from '@ethersproject/providers'
+import { getInjectedProvider, subscribeToInjectedProvider } from '../../helpers/injectedProvider'
+import { useNetwork } from '../network'
+import { useConfig } from '../config'
+
+interface InjectedProviderProviderProps {
+  children: ReactNode
+}
+
+export function InjectedProviderProvider({ children }: InjectedProviderProviderProps) {
+  const { update, reportError } = useNetwork()
+  const { pollingInterval } = useConfig()
+  const [injectedProvider, setInjectedProvider] = useState<Web3Provider | undefined>()
+
+  useEffect(function () {
+    getInjectedProvider(pollingInterval).then(setInjectedProvider)
+  }, [])
+
+  useEffect(() => {
+    const underlyingProvider: any = injectedProvider?.provider
+    return subscribeToInjectedProvider(underlyingProvider, update, reportError)
+  }, [injectedProvider])
```

Comment:
Looks like eager subscription, shouldn't we only subscribe one we trigger `connect`?

Reply:
This mimics the current behaviour. I agree this can be subject to change but let's move the discussion to a different PR

Reply:
👍

Reply:
BTW we will need to subscribe to same set of events when implementing foreign providers support, this will be a good place to refactor this